python/ray/util/spark/databricks_hook.py

```diff
@@ -54,21 +58,101 @@ def display_databricks_driver_proxy_url(spark_context, port, title):
 
     def on_ray_dashboard_created(self, port):
         display_databricks_driver_proxy_url(
             get_spark_session().sparkContext, port, "Ray Cluster Dashboard"
         )
 
-    def on_spark_background_job_created(self, job_group_id):
+    def on_cluster_created(self, ray_cluster_handler):
+        db_api_entry = _get_db_api_entry()
+        try:
+            db_api_entry.registerBackgroundSparkJobGroup(
+                ray_cluster_handler.spark_job_group_id
+            )
+        except Exception:
+            _logger.warning(
+                "Registering ray cluster spark job as background job failed. "
+                "You need to manually call `ray.util.spark.shutdown_ray_cluster()` "
+                "before detaching your databricks notebook."
+            )
+
+        auto_shutdown_timeout_minutes = float(
+            os.environ.get(DATABRICKS_RAY_ON_SPARK_AUTOSHUTDOWN_TIMEOUT_MINUTES, "30")
+        )
+        if auto_shutdown_timeout_minutes == 0:
+            _logger.info(
+                "The Ray cluster will keep running until you manually detach the "
+                "databricks notebook or call "
+                "`ray.util.spark.shutdown_ray_cluster()`."
+            )
+        elif auto_shutdown_timeout_minutes < 0:
+            raise ValueError(
+                "You must set "
+                f"'{DATABRICKS_RAY_ON_SPARK_AUTOSHUTDOWN_TIMEOUT_MINUTES}' "
+                "to a value >= 0."
+            )
+        else:
+            _logger.info(
+                "The Ray cluster will be shut down automatically if you don't run "
+                "commands on the databricks notebook for "
+                f"{auto_shutdown_timeout_minutes} minutes. You can change the "
+                "timeout minutes by setting "
+                "'DATABRICKS_RAY_ON_SPARK_AUTOSHUTDOWN_TIMEOUT_MINUTES' environment "
+                "variable, setting it to 0 means infinite timeout."
+            )
+
+        def auto_shutdown_watcher():
+            auto_shutdown_timeout_millis = auto_shutdown_timeout_minutes * 60 * 1000
+            while True:
+                if ray_cluster_handler.is_shutdown:
+                    # The cluster is shut down. The watcher thread exits.
+                    return
+
+                idle_time = db_api_entry.getIdleTimeMillisSinceLastNotebookExecution()
+
+                if idle_time > auto_shutdown_timeout_millis:
+                    from ray.util.spark import cluster_init
+
+                    with cluster_init._active_ray_cluster_rwlock:
+                        if ray_cluster_handler is cluster_init._active_ray_cluster:
+                            cluster_init.shutdown_ray_cluster()
+                    return
+
+                time.sleep(DATABRICKS_AUTO_SHUTDOWN_POLL_INTERVAL_SECONDS)
+
         try:
-            get_dbutils().entry_point.registerBackgroundSparkJobGroup(job_group_id)
+            db_api_entry.getIdleTimeMillisSinceLastNotebookExecution()
```

Comment:
We can move this block before `auto_shutdown_watcher`. If `getIdleTimeMillisSinceLastNotebookExecution` is unavailable, it's useless to define `auto_shutdown_watcher`, right?